MPSGameLift/Code/Source/RegionalLatencySystemComponent.cpp

```diff
@@ -0,0 +1,102 @@
+    {
+        AZ::Interface<IRegionalLatencyFinder>::Register(this);
+    }
+
+    void RegionalLatencySystemComponent::Deactivate()
+    {
+        AZ::Interface<IRegionalLatencyFinder>::Unregister(this);
+    }
+
+    void RegionalLatencySystemComponent::Reflect(AZ::ReflectContext* context)
+    {
+        if (const auto serializeContext = azrtti_cast<AZ::SerializeContext*>(context))
+        {
+            serializeContext->Class<RegionalLatencySystemComponent, AZ::Component>()
+                ->Version(1)
+            ;
+        }
+    }
+
+    void RegionalLatencySystemComponent::RequestLatencies()
+    {
+        // Response callbacks are received on a separate thread. 
+        // Users can call RequestLatencies from main thread multiple times before all the requests
+        // have returned a response. For simplicity, don't make more requests if requests are still waiting for a response.
+        if (m_responsesPending.load() > 0)
+        {
+            AZ_Warning("RegionalLatencySystemComponent", false, 
+                "Denying RequestLatencies. "
+                "Latencies are already being received; "
+                "listen for RegionalLatencyFinderNotifications::OnRequestLatenciesComplete to get a notification once all region latencies have been received.");
+            return;
+        }
+
+        // Start pinging region endpoints
+        m_responsesPending.store(aznumeric_cast<int>(AZStd::size(Regions)));
+
+        for (auto region : Regions)
+        {
+            AZStd::string regionEndpoint = AZStd::string::format(RegionalEndpointUrlFormat, region);
+
+            HttpRequestor::HttpRequestorRequestBus::Broadcast(&HttpRequestor::HttpRequestorRequests::AddTextRequest, regionEndpoint, Aws::Http::HttpMethod::HTTP_GET,
+                [this, region]([[maybe_unused]]const AZStd::string& response, Aws::Http::HttpResponseCode responseCode)
+                {
+                    if (responseCode == Aws::Http::HttpResponseCode::OK)
+                    {
+                        AZStd::chrono::milliseconds roundTripTime;
+                        HttpRequestor::HttpRequestorRequestBus::BroadcastResult(roundTripTime, &HttpRequestor::HttpRequestorRequests::GetLastRoundTripTime);
+
+                        AZStd::lock_guard<AZStd::mutex> lock(m_mapMutex);
+                        m_regionalLatencies[region] = roundTripTime;
+                    }
+                    else
+                    {
+                        AZ_Error("RegionalLatencySystemComponent", false, "Failed to receive response for region: %s.", region);
+                    }
+
+                    // Check if all requests have responded
+                    m_responsesPending.fetch_sub(1);
+                    if (m_responsesPending.load() == 0)
+                    {
+                        RegionalLatencyFinderNotificationBus::Broadcast(&RegionalLatencyFinderNotifications::OnRequestLatenciesComplete, m_regionalLatencies);
+                    }
```

Comment:
suggestion, rather than using a lock on `m_regionalLatencies` use a temporary array passed as a local into the callback, and then on line 81 after you get all the responses, use AZ::TickBus::Enqueue to enqueue a lambda that sets the m_responsesPending map on the main thread and then does the broadcast after.  That way you won't need the mutex and whoever receives the notification doesn't have to worry about receiving a notification from thread that isn't the main thread.